Bug 1814397: fix wrongful backup of files not originally on the system

yuqi-zhang opened the pull request:
Adds to https://github.com/openshift/machine-config-operator/pull/1586

Add a check for backwards compatibility: basically if the file
doesn't exist in /usr/etc/ and no rpm is claming it, we assume
that the orig file came from a wrongful backup of a MachineConfig
file instead of a RHCOS file.

runcom commented on pkg/daemon/update.go:
```diff
@@ -788,6 +788,25 @@ func (dn *Daemon) deleteStaleData(oldConfig, newConfig *mcfgv1.MachineConfig) er
+				// and no rpm is claming it, we assume that the orig file came from a wrongful backup of a
+				// MachineConfig file instead of a RHCOS file. See https://bugzilla.redhat.com/show_bug.cgi?id=1814397
+				if strings.HasPrefix(origFileName(f.Path), "/etc") {
+					if _, err := os.Stat("/usr" + origFileName(f.Path)); err != nil {
```
should this be just `f.Path`? `/usr/etc` contains paths _in_ `/etc` whether `origFileName` turns `f.Path` into `/etc/machine-config-daemon/orig/f.Name` which won't live by default under `/usr/etc` which stores only RHCOS default

runcom commented on pkg/daemon/update.go:
```diff
@@ -788,6 +788,25 @@ func (dn *Daemon) deleteStaleData(oldConfig, newConfig *mcfgv1.MachineConfig) er
+					if err := os.Remove(origFileName(f.Path)); err != nil {
+						return errors.Wrapf(err, "deleting orig file %q: %v", origFileName(f.Path), err)
+					}
+					continue
```
instead of continuing, should we go ahead at the routine that deletes the file (at https://github.com/openshift/machine-config-operator/pull/1593/commits/7a68edd2d2b87ea481fde86f7c3a5c849565bea4#diff-06961b075f1753956d802ba954d2cfb5R819)

runcom commented on pkg/daemon/update.go:
```diff
@@ -788,6 +788,25 @@ func (dn *Daemon) deleteStaleData(oldConfig, newConfig *mcfgv1.MachineConfig) er
+						if err := os.Remove(origFileName(f.Path)); err != nil {
+							return errors.Wrapf(err, "deleting orig file %q: %v", origFileName(f.Path), err)
+						}
+						continue
```
same q as https://github.com/openshift/machine-config-operator/pull/1593/commits/7a68edd2d2b87ea481fde86f7c3a5c849565bea4#r399142540

runcom commented:
manually upgraded from 4.2 to 4.4 (with this patch) and it worked just fine, I'm gonna test all the way from 4.1 to fully verify this, otherwise, this replaces https://github.com/openshift/machine-config-operator/pull/1586

/approve

runcom commented:
The additional patches here are also backported already to https://github.com/openshift/machine-config-operator/pull/1588 - we'd need a 4.3 patch and 4.2 patch anyway but those can wait till we unblock 4.4 as this patch effectively fixes the upgrade.

runcom commented:
/retest

yuqi-zhang commented:
Ah, thanks for the fix! Feel free to squash the commits together

kikisdeliveryservice commented on pkg/daemon/update.go:
```diff
@@ -782,12 +782,35 @@ func (dn *Daemon) deleteStaleData(oldConfig, newConfig *mcfgv1.MachineConfig) er
-					return err
+			if _, err := os.Stat(noOrigFileName(f.Path)); err == nil {
+				if err := os.Remove(noOrigFileName(f.Path)); err != nil {
+					return errors.Wrapf(err, "deleting no orig file %q: %v", noOrigFileName(f.Path), err)
```
this error message seems confusing to a user

kikisdeliveryservice replied:
To be clear I'm having trouble understanding what this noOrigFile.... stuff is doing.

yuqi-zhang replied:
`noOrig` just means that there was no original file on disk when we tried to write a file via the MCD. This tells the MCD to not revert a file when the last instance of it is deleted in MachineConfigs.

The reasoning is added to `createOrigFile` function. It's also in the commit message. Where else would you like me to add some comments? Or how else would you like it to be named

kikisdeliveryservice replied:
maybe noOrigFileFlag or noOrigFileStamp?

kikisdeliveryservice commented on pkg/daemon/update.go:
```diff
@@ -997,16 +1020,32 @@ func origParentDir() string {
 	return filepath.Join(origParentDir(), fpath+".mcdorig")
 }
 
+func noOrigFileName(fpath string) string {
```
i dont really understand this naming?  Can we make it clearer since it's not really clear what's going on here... and the naming of this noOrig doesn't really make it clear

kikisdeliveryservice replied:
ie anyone who will have to maintain this in the future is going to have trouble understanding what this is for...

yuqi-zhang replied:
I can add some comments above this as well. To be clear this is to unblock upgrades for 4.4 and not the most clean long term solution. We are attempting to fix an insidious error that we didn't catch way back

yuqi-zhang commented:
Pushed new commit:
Squashed fixes
Added comments and changed naming for clarity
Add new fix for files that exist in `/etc`

Manually tested to be good for both new and existing scenarios

kikisdeliveryservice commented on pkg/daemon/update.go:
```diff
@@ -782,12 +782,33 @@ func (dn *Daemon) deleteStaleData(oldConfig, newConfig *mcfgv1.MachineConfig) er
-					return err
+			if _, err := os.Stat(noOrigFileStampName(f.Path)); err == nil {
+				if err := os.Remove(noOrigFileStampName(f.Path)); err != nil {
+					return errors.Wrapf(err, "deleting no orig file %q: %v", noOrigFileStampName(f.Path), err)
```
could you update error message here?

yuqi-zhang commented:
Modified error msg

Now I think about it technically I think the logic is wrong.

/hold

yuqi-zhang commented:
/hold cancel

After some more thought I think this is alright. To be clear if we have an `.orig` file, the conditions it will restore is that it must both be a) managed by an RPM and b) exist in the original ostree (if it was in `/etc`).

Is there a scenario where we have an rpm drop in a file in `/etc` but it would not exist in `/usr/etc`? Conversely, is there a scenario where it exists in `/usr/etc` but no rpm is claiming it?

runcom commented:
> Is there a scenario where we have an rpm drop in a file in `/etc` but it would not exist in `/usr/etc`? 

I think this is only possible if you install an rpm after initial bootstrap + MCO right?

> Conversely, is there a scenario where it exists in `/usr/etc` but no rpm is claiming it?

I think yes, say you installed an RPM with a spec that doesn't claim ownership of a file (it happens a lot, trust me...) but still ships the file under `/etc`

runcom commented:
/retest

yuqi-zhang commented:
I can confirm this passes the tests to be added in https://github.com/openshift/machine-config-operator/pull/1590

I can also add a test to ensure existing `.orig`s are properly pruned as a separate PR. For now I have done manual testing and can confirm this should fix those cases.

runcom commented:
Tested once more an upgrade from 4.2 up to this PR in 4.4 (https://github.com/openshift/machine-config-operator/pull/1588) and the upgrade succeded and all files were correctly pruned.

/lgtm

sinnykumari commented on pkg/daemon/update.go:
```diff
@@ -782,20 +782,33 @@ func (dn *Daemon) deleteStaleData(oldConfig, newConfig *mcfgv1.MachineConfig) er
+					if err := os.Remove(origFileName(f.Path)); err != nil {
+						return errors.Wrapf(err, "deleting orig file %q: %v", origFileName(f.Path), err)
+					}
+				} else if _, err := os.Stat("/usr" + f.Path); strings.HasPrefix(f.Path, "/etc") && os.IsNotExist(err) {
```
I think we should keep /usr/etc/ file content checks only for OSTree based system. On a node containing traditional OS, `else` section would never run because it doesn't have etc content in /usr/etc/ as well.

runcom replied:
right, tho, on normal system /usr/etc isn't containing the base /etc - I agree tho that we need a check just in case of RHCOS-ostree nodes to avoid RHEL hosts misbehaving (remote option but still)

runcom commented:
/hold

pending Sinny's review

sinnykumari commented on pkg/daemon/update.go:
```diff
@@ -997,16 +1018,39 @@ func origParentDir() string {
 	return filepath.Join("/etc", "machine-config-daemon", "orig")
 }
 
+func noOrigParentDir() string {
```
It will be nice to have description of what this function does.

sinnykumari commented on pkg/daemon/update.go:
```diff
@@ -782,12 +782,33 @@ func (dn *Daemon) deleteStaleData(oldConfig, newConfig *mcfgv1.MachineConfig) er
+				glog.V(2).Infof("Removing file %q completely", f.Path)
+			} else if _, err := os.Stat(origFileName(f.Path)); err == nil {
+				// Add a check for backwards compatibility: basically if the file doesn't exist in /usr/etc
+				// and no rpm is claming it, we assume that the orig file came from a wrongful backup of a
```
typo s/claming/claiming/

sinnykumari commented on pkg/daemon/update.go:
```diff
@@ -782,12 +782,33 @@ func (dn *Daemon) deleteStaleData(oldConfig, newConfig *mcfgv1.MachineConfig) er
+			} else if _, err := os.Stat(origFileName(f.Path)); err == nil {
+				// Add a check for backwards compatibility: basically if the file doesn't exist in /usr/etc
+				// and no rpm is claming it, we assume that the orig file came from a wrongful backup of a
+				// MachineConfig file instead of a RHCOS file. See https://bugzilla.redhat.com/show_bug.cgi?id=1814397
```
Since [FCOS support has been added](https://github.com/openshift/machine-config-operator/pull/1496), saying CoreOS instead of RHCOS would be more generic

sinnykumari commented on pkg/daemon/update.go:
```diff
@@ -997,16 +1018,39 @@ func origParentDir() string {
 		return nil
 	}
+	if _, err := os.Stat(fpath); os.IsNotExist(err) {
+		// create a noorig file that tells the MCD that the file wasn't present on disk in RHCOS
```
similar to previous comment
s/RHCOS/CoreOS

yuqi-zhang commented:
Added extra check per comment and fixed up some typos

sinnykumari commented:
Looks sane to me
/hold cancel
/approve

runcom commented:
/retest

yuqi-zhang commented:
/retest

yuqi-zhang commented:
/retest

yuqi-zhang commented:
/retest

sinnykumari commented:
/test e2e-aws

sinnykumari commented:
/test e2e-aws

sinnykumari commented:
/retest

runcom commented:
/lgtm
/retest

runcom commented:
uhm, wondering if something broke the install but weird it's on for e2e-aws

runcom commented:
/refresh

runcom commented:
/cherry-pick release-4.4

runcom commented:
/cherry-pick release-4.3

vrutkovs commented:
/cherrypick fcos
